proposals/0052-fonts.md

```diff
@@ -0,0 +1,369 @@
+
+# Summary
+
+Have first-party support for fonts in Astro.
+
+# Example
+
+```js
+// astro config
+export default defineConfig({
+  fonts: {
+    families: ["Roboto", "Lato"],
+  },
+});
+```
+
+```astro
+---
+// layouts/Layout.astro
+import { Font } from 'astro:fonts'
+---
+<head>
+	<Font family='Inter' preload />
+	<Font family='Lato' />
+	<style>
+		h1 {
+			font-family: var(--astro-font-inter);
+		}
+		p {
+			font-family: var(--astro-font-lato);
+		}
+	</style>
+</head>
+```
+
+# Background & Motivation
+
+Fonts is one of those basic things when making a website, but also an annoying one to deal with. Should I just use a link to a remote font? Or download it locally? How should I handle preloads then?
+
+The goal is to improve the DX around using fonts in Astro.
+
+> Why not using fontsource?
+
+Fontsource is great! But it's not intuitive to preload, and more importantly, doesn't have all fonts. The goal is to have a more generic API for fonts (eg. you want to use a paid provider like adobe).
+
+# Goals
+
+- Specify what font to use
+- Cache fonts
+- Specify what provider to use
+- Load/preload font on a font basis
+- Generate fallbacks automatically
+- Performant defaults
+- Runtime agnostic
+- Configure font families (subset, unicode range, weights etc)
+
+# Non-Goals
+
+- Runtime API (SSR is supported tho)
+- Automatic subsetting (eg. analyzing static content)
+- Font detection
+
+# Detailed Design
+
+## Astro config
+
+### Overview
+
+The goal is to have a config that starts really simple for basic usecases, but can also be complex for advanced usecases. Here's an example of basic config:
+
+```js
+import { defineConfig } from "astro/config";
+
+export default defineConfig({
+  fonts: {
+    families: ["Roboto", "Lato"],
+  },
+});
+```
+
+That would get fonts from [Google Fonts](https://fonts.google.com/) with sensible defaults.
+
+Here's a more complex example:
+
+```js
+import { defineConfig, fontProviders } from "astro/config";
+import { myCustomFontProvider } from "./provider";
+
+export default defineConfig({
+  fonts: {
+    providers: [
+      fontProviders.adobe({ apiKey: process.env.ADOBE_FONTS_API_KEY }),
+      myCustomFontProvider(),
+    ],
+    defaults: {
+      provider: "adobe",
+      weights: [200, 700],
+      styles: ["italic"],
+      subsets: [
+        "cyrillic-ext",
+        "cyrillic",
+        "greek-ext",
+        "greek",
+        "vietnamese",
+        "latin-ext",
+        "latin",
+      ],
+    },
+    families: [
+      "Roboto",
+      {
+        name: "Lato",
+        provider: "google",
+        weights: [100, 200, 300],
+      },
+      {
+        name: "Custom",
+        provider: "local",
+        src: ["./assets/fonts/Custom.woff2"],
+      },
+    ],
+  },
+});
+```
+
+### Providers
+
+#### Definition
+
+A provider allows to retrieve font faces data from a font family name from a given CDN or abstraction. It's a [unifont](https://github.com/unjs/unifont) provider.
```

Comment:
I think some people could have a concern with us exposing a third-party "API" like that in the config, but we already do for syntax highlighting themes, so I think it's okay. Plus, the default providers unifont provide are probably enough for 99.999% of people anyway.